src/components/ReadingDots.jsx

```diff
@@ -0,0 +1,100 @@
+import React, { useState, useEffect, useRef } from 'react';
+
+// Generate a unique and stable hue based on the actor name
+function getActorHue(actor) {
+  let input = 0;
+  for (const char of actor) {
+    input += char.charCodeAt(0);
+  }
+
+  return input % 360;
+}
+
+const ReadingDots = ({ actors, currentActor, currentScript }) => {
+  const [dots, setDots] = useState({});
+  const refDots = useRef({});
+  refDots.current = dots;
+
+  const actorNameMap = { };
+  for (const actor of actors) {
+    actorNameMap[actor.id] = actor.name;
+  }
+
+  useEffect(() => {
+    // get initial dots positions
+    const source = new EventSource(`/positions/${currentScript}`);
+    const positionReceiver = function (message) {
+      let m = JSON.parse(message.data);
+
+      const newDots = Object.assign({}, refDots.current);
+
+      newDots[m.actorId] = m.position;
+
+      setDots(newDots);
+      refDots.current = newDots;
+    };
+
+    source.addEventListener('message', positionReceiver);
+
+    let scrollTimeout;
+
+    let positionReporter = () => {
+      // replace a currently queued update if we're still scrolling before the update was sent.
+      // this helps reduce many small changes into one larger change request
+      clearTimeout(scrollTimeout);
+
+      scrollTimeout = setTimeout(() => {
+        let readingDotsHeightOffset =
+          document.getElementsByTagName('body')[0].getBoundingClientRect().height -
+          document.getElementById('readingDots').getBoundingClientRect().height;
+
+        let position = (window.scrollY - readingDotsHeightOffset) / (document.body.scrollHeight - readingDotsHeightOffset);
+        position = position.toPrecision(4);
+
+        fetch('/positions', {
```

Comment:
Does this need to be asynchronous as well or is that handled differently since it's an open eventSource instead of something that closes off after each request?

Reply:
We could await this, but I don't think it's necessary since we don't have any code we need to run after or any try catch blocks around it